File: tiny-event-sourcing-spring-app/src/main/kotlin/ru/quipy/shop/product/Product.java
package ru.quipy.shop.product;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;
import ru.quipy.core.annotations.StateTransitionFunc;
import ru.quipy.shop.payment.entities.PaymentStatus;
import ru.quipy.shop.payment.events.PaymentChangeStatusEvent;
import ru.quipy.shop.product.events.ProductChangePrice;
import ru.quipy.shop.product.events.ProductCreateEvent;

import java.util.UUID;

@Getter
@AllArgsConstructor
public class Product {
    private UUID id;
    private final String name;
    @Setter
    private Long price;

    public UUID getId() {
        return id;
    }

    @StateTransitionFunc
    public void createProduct(ProductCreateEvent event) {
        id = event.getProductId();
    }

    @StateTransitionFunc
    public void changePrice(ProductChangePrice event) {
        price = event.getPrice();
    }

    public ProductChangePrice changeStatus(Long price) {
        return new ProductChangePrice(id, price);
    }

    public ProductCreateEvent createProduct(Long totalPrice) {
        return new ProductCreateEvent(id, name, price);
    }
}

File: tiny-event-sourcing-spring-app/src/main/kotlin/ru/quipy/shop/product/Product.kt
package ru.quipy.shop.product

import lombok.AllArgsConstructor
import lombok.Getter
import lombok.Setter
import ru.quipy.core.annotations.StateTransitionFunc
import ru.quipy.domain.AggregateState
import ru.quipy.shop.product.events.ProductChangePrice
import ru.quipy.shop.product.events.ProductCreateEvent

import java.util.UUID

@Getter
@AllArgsConstructor
class Product : AggregateState<UUID, ProductAggregate> {
    private lateinit var id: UUID
    lateinit var name: String
    @Setter
    var price: Long = 0
        get() = price


    @StateTransitionFunc
    fun createProduct(event: ProductCreateEvent) {
        id = event.productId
    }

    @StateTransitionFunc
    fun changePrice(event: ProductChangePrice) {
        price = event.price
    }

    fun changePrice(price: Long): ProductChangePrice =
        ProductChangePrice(id, price)

    fun createProduct(id: UUID, name: String, totalPrice: Long): ProductCreateEvent =
        ProductCreateEvent(id, name, price)

    override fun getId(): UUID? = id
}
